File: internal/plan/plan_test.go
package plan

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"google.golang.org/protobuf/types/known/timestamppb"

	storepb "github.com/usememos/memos/proto/gen/store"
)

func TestEffective(t *testing.T) {
	now := time.Date(2026, 9, 23, 12, 0, 0, 0, time.UTC)
	teams := func(expire *timestamppb.Timestamp) *storepb.PackageUserSetting {
		return &storepb.PackageUserSetting{Plan: storepb.PackageUserSetting_TEAMS, ExpireTime: expire}
	}

	tests := []struct {
		name string
		pkg  *storepb.PackageUserSetting
		want Tier
	}{
		{"no package", nil, TierFree},
		{"free package", &storepb.PackageUserSetting{Plan: storepb.PackageUserSetting_FREE}, TierFree},
		{"unspecified plan", &storepb.PackageUserSetting{}, TierFree},
		{"teams without expiry", teams(nil), TierTeams},
		{"teams before expiry", teams(timestamppb.New(now.Add(time.Hour))), TierTeams},
		{"teams at expiry", teams(timestamppb.New(now)), TierFree},
		{"teams after expiry", teams(timestamppb.New(now.Add(-time.Hour))), TierFree},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			assert.Equal(t, tt.want, Effective(tt.pkg, now))
		})
	}
}

File: web/tests/plan.test.ts
import { create } from "@bufbuild/protobuf";
import { timestampFromDate } from "@bufbuild/protobuf/wkt";
import { describe, expect, it } from "vitest";
import { UserSetting_PackageSetting_Plan, UserSetting_PackageSettingSchema } from "@/types/proto/api/v1/user_service_pb";
import { isTeamsActive } from "@/utils/plan";

const now = new Date("2026-09-23T12:00:00Z");
const teams = (expire?: Date) =>
  create(UserSetting_PackageSettingSchema, {
    plan: UserSetting_PackageSetting_Plan.TEAMS,
    expireTime: expire ? timestampFromDate(expire) : undefined,
  });

describe("isTeamsActive", () => {
  it("is false without a package or on Free", () => {
    expect(isTeamsActive(undefined, now)).toBe(false);
    expect(isTeamsActive(create(UserSetting_PackageSettingSchema, { plan: UserSetting_PackageSetting_Plan.FREE }), now)).toBe(false);
  });

  it("is true for Teams without an expiry or before it", () => {
    expect(isTeamsActive(teams(), now)).toBe(true);
    expect(isTeamsActive(teams(new Date("2026-09-24T12:00:00Z")), now)).toBe(true);
  });

  it("is false once the expiry has passed", () => {
    expect(isTeamsActive(teams(now), now)).toBe(false);
    expect(isTeamsActive(teams(new Date("2026-09-22T12:00:00Z")), now)).toBe(false);
  });
});
